control-plane-operator/controllers/awsprivatelink/awsprivatelink_controller.go

```diff
@@ -299,51 +316,108 @@ func reconcileAWSEndpointService(ctx context.Context, awsEndpointService *hyperv
 			return fmt.Errorf("endpoint %s not found, resetting status", serviceName)
 		}
 		log.Info("endpoint exists", "endpointID", endpointID)
-		return nil
+		endpointDNSEntries = output.VpcEndpoints[0].DnsEntries
+	} else {
+		if hcp.Spec.Platform.Type != hyperv1.AWSPlatform || hcp.Spec.Platform.AWS == nil || hcp.Spec.Platform.AWS.CloudProviderConfig == nil {
+			return fmt.Errorf("AWS platform information not provided in HostedControlPlane")
+		}
+		input := &ec2.CreateVpcEndpointInput{
+			ServiceName:     aws.String(serviceName),
+			VpcId:           aws.String(hcp.Spec.Platform.AWS.CloudProviderConfig.VPC),
+			VpcEndpointType: aws.String(ec2.VpcEndpointTypeInterface),
+		}
+		if hcp.Spec.Platform.AWS.CloudProviderConfig.Subnet != nil && hcp.Spec.Platform.AWS.CloudProviderConfig.Subnet.ID != nil {
+			input.SubnetIds = []*string{hcp.Spec.Platform.AWS.CloudProviderConfig.Subnet.ID}
+		}
+		output, err := ec2Client.CreateVpcEndpointWithContext(ctx, input)
+		if err != nil {
+			return err
+		}
+		if output == nil || output.VpcEndpoint == nil {
+			return fmt.Errorf("CreateVpcEndpointWithContext output is nil")
+		}
 
+		endpointID = *output.VpcEndpoint.VpcEndpointId
+		log.Info("endpoint created", "endpointID", endpointID)
+		awsEndpointService.Status.EndpointID = endpointID
+		endpointDNSEntries = output.VpcEndpoint.DnsEntries
 	}
 
-	if hcp.Spec.Platform.Type != hyperv1.AWSPlatform || hcp.Spec.Platform.AWS == nil || hcp.Spec.Platform.AWS.CloudProviderConfig == nil {
-		return fmt.Errorf("AWS platform information not provided in HostedControlPlane")
-	}
-	input := &ec2.CreateVpcEndpointInput{
-		ServiceName:     aws.String(serviceName),
-		VpcId:           aws.String(hcp.Spec.Platform.AWS.CloudProviderConfig.VPC),
-		VpcEndpointType: aws.String(ec2.VpcEndpointTypeInterface),
+	if len(endpointDNSEntries) == 0 {
+		log.Info("endpoint has no DNS entries, skipping DNS record creation", "endpointID", endpointID)
+		return nil
 	}
-	if hcp.Spec.Platform.AWS.CloudProviderConfig.Subnet != nil && hcp.Spec.Platform.AWS.CloudProviderConfig.Subnet.ID != nil {
-		input.SubnetIds = []*string{hcp.Spec.Platform.AWS.CloudProviderConfig.Subnet.ID}
+
+	var recordName string
+	if awsEndpointService.Name == "kube-apiserver-private" {
+		recordName = "api"
+	} else if strings.HasPrefix(awsEndpointService.Name, "router-") {
+		recordName = "*.apps"
+	} else {
+		return fmt.Errorf("no mapping from AWSEndpointService to DNS")
 	}
-	output, err := ec2Client.CreateVpcEndpointWithContext(ctx, input)
+
+	zoneName := fmt.Sprintf("%s.%s", hcp.Name, hypershiftLocalZone)
+	zoneID, err := lookupZoneID(ctx, route53Client, zoneName)
 	if err != nil {
 		return err
 	}
 
-	endpointID = *output.VpcEndpoint.VpcEndpointId
-	log.Info("endpoint created", "endpointID", endpointID)
-	awsEndpointService.Status.EndpointID = endpointID
+	fqdn := fmt.Sprintf("%s.%s", recordName, zoneName)
+	if fqdn != "" && zoneID != "" {
```

Comment:
fqdn will always be !="", even if the variables are empty fqdn will be "." right?

Reply:
Good catch.  There are some unnecessary checks there.